File: projectforge-business/src/main/java/org/projectforge/business/orga/ContractDO.java
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2014 Kai Reinhard (k.reinhard@micromata.de)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.orga;

import org.hibernate.search.annotations.Analyze;
import org.hibernate.search.annotations.DateBridge;
import org.hibernate.search.annotations.EncodingType;
import org.hibernate.search.annotations.Field;
import org.hibernate.search.annotations.FieldBridge;
import org.hibernate.search.annotations.Index;
import org.hibernate.search.annotations.Indexed;
import org.hibernate.search.annotations.Resolution;
import org.hibernate.search.annotations.Store;
import org.hibernate.search.bridge.builtin.IntegerBridge;
import org.projectforge.framework.persistence.entities.DefaultBaseDO;

import java.sql.Date;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;

/**
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
@Entity
@Indexed
@Table(name = "T_CONTRACT", uniqueConstraints = {
    @UniqueConstraint(columnNames = { "number", "tenant_id" }) }, indexes = {
    @javax.persistence.Index(name = "idx_fk_t_contract_tenant_id", columnList = "tenant_id")
})
public class ContractDO extends DefaultBaseDO
{
  private static final long serialVersionUID = -1399338188515793833L;

  @Field(index = Index.YES, analyze = Analyze.NO /* UN_TOKENIZED */, store = Store.NO,
      bridge = @FieldBridge(impl = IntegerBridge.class))
  private Integer number;

  @Field(index = Index.YES, analyze = Analyze.NO /* UN_TOKENIZED */)
  @DateBridge(resolution = Resolution.DAY, encoding = EncodingType.STRING)
  private Date date;

  @Field(index = Index.YES, analyze = Analyze.NO /* UN_TOKENIZED */)
  @DateBridge(resolution = Resolution.DAY, encoding = EncodingType.STRING)
  private Date validFrom;

  @Field(index = Index.YES, analyze = Analyze.NO /* UN_TOKENIZED */)
  @DateBridge(resolution = Resolution.DAY, encoding = EncodingType.STRING)
  private Date validUntil;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String title;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String coContractorA;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String contractPersonA;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String signerA;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String coContractorB;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String contractPersonB;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String signerB;

  @Field(index = Index.YES, analyze = Analyze.NO /* UN_TOKENIZED */)
  @DateBridge(resolution = Resolution.DAY, encoding = EncodingType.STRING)
  private Date signingDate;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String type;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String status;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String text;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String reference;

  @Field(index = Index.YES /* TOKENIZED */, store = Store.NO)
  private String filing;

  @Field(index = Index.YES, analyze = Analyze.NO /* UN_TOKENIZED */)
  @DateBridge(resolution = Resolution.DAY, encoding = EncodingType.STRING)
  private Date resubmissionOnDate;

  @Field(index = Index.YES, analyze = Analyze.NO /* UN_TOKENIZED */)
  @DateBridge(resolution = Resolution.DAY, encoding = EncodingType.STRING)
  private Date dueDate;

  /**
   * consecutively numbered.
   */
  @Column(nullable = true)
  public Integer getNumber()
  {
    return number;
  }

  public ContractDO setNumber(final Integer number)
  {
    this.number = number;
    return this;
  }

  @Column(name = "c_date")
  public Date getDate()
  {
    return date;
  }

  public ContractDO setDate(final Date date)
  {
    this.date = date;
    return this;
  }

  /**
   * @return Start of the validity period.
   */
  @Column(name = "valid_from")
  public Date getValidFrom()
  {
    return validFrom;
  }

  public ContractDO setValidFrom(final Date validFrom)
  {
    this.validFrom = validFrom;
    return this;
  }

  /**
   * @return End of the validity period.
   */
  @Column(name = "valid_until")
  public Date getValidUntil()
  {
    return validUntil;
  }

  public ContractDO setValidUntil(final Date validUntil)
  {
    this.validUntil = validUntil;
    return this;
  }

  @Column(length = 1000)
  public String getTitle()
  {
    return title;
  }

  public ContractDO setTitle(final String title)
  {
    this.title = title;
    return this;
  }

  /**
   * Types (as free texts) are configurable in ProjectForge's config file.
   */
  @Column(length = 100)
  public String getType()
  {
    return type;
  }

  public ContractDO setType(final String type)
  {
    this.type = type;
    return this;
  }

  @Column(length = 100)
  public String getStatus()
  {
    return status;
  }

  public ContractDO setStatus(final String status)
  {
    this.status = status;
    return this;
  }

  @Column(length = 4000)
  public String getText()
  {
    return text;
  }

  public ContractDO setText(final String text)
  {
    this.text = text;
    return this;
  }

  @Column(length = 1000)
  public String getReference()
  {
    return reference;
  }

  public ContractDO setReference(final String reference)
  {
    this.reference = reference;
    return this;
  }

  @Column(name = "resubmission_on_date")
  public Date getResubmissionOnDate()
  {
    return resubmissionOnDate;
  }

  public ContractDO setResubmissionOnDate(final Date resubmissionOnDate)
  {
    this.resubmissionOnDate = resubmissionOnDate;
    return this;
  }

  @Column(name = "due_date")
  public Date getDueDate()
  {
    return dueDate;
  }

  public ContractDO setDueDate(final Date dueDate)
  {
    this.dueDate = dueDate;
    return this;
  }

  @Column(length = 1000)
  public String getFiling()
  {
    return filing;
  }

  public ContractDO setFiling(final String filing)
  {
    this.filing = filing;
    return this;
  }

  @Column(length = 1000, name = "co_contractor_a")
  public String getCoContractorA()
  {
    return coContractorA;
  }

  public ContractDO setCoContractorA(final String coContractorA)
  {
    this.coContractorA = coContractorA;
    return this;
  }

  @Column(length = 1000, name = "contract_person_a")
  public String getContractPersonA()
  {
    return contractPersonA;
  }

  public ContractDO setContractPersonA(final String contractPersonA)
  {
    this.contractPersonA = contractPersonA;
    return this;
  }

  @Column(length = 1000, name = "signer_a")
  public String getSignerA()
  {
    return signerA;
  }

  public ContractDO setSignerA(final String signerA)
  {
    this.signerA = signerA;
    return this;
  }

  @Column(length = 1000, name = "co_contractor_b")
  public String getCoContractorB()
  {
    return coContractorB;
  }

  public ContractDO setCoContractorB(final String coContractorB)
  {
    this.coContractorB = coContractorB;
    return this;
  }

  @Column(length = 1000, name = "contract_person_b")
  public String getContractPersonB()
  {
    return contractPersonB;
  }

  public ContractDO setContractPersonB(final String contractPersonB)
  {
    this.contractPersonB = contractPersonB;
    return this;
  }

  @Column(length = 1000, name = "signer_b")
  public String getSignerB()
  {
    return signerB;
  }

  public ContractDO setSignerB(final String signerB)
  {
    this.signerB = signerB;
    return this;
  }

  @Column(name = "signing_date")
  public Date getSigningDate()
  {
    return signingDate;
  }

  public ContractDO setSigningDate(final Date signingDate)
  {
    this.signingDate = signingDate;
    return this;
  }
}

File: projectforge-business/src/main/kotlin/org/projectforge/business/orga/ContractDO.kt
/////////////////////////////////////////////////////////////////////////////
//
// Project ProjectForge Community Edition
//         www.projectforge.org
//
// Copyright (C) 2001-2024 Micromata GmbH, Germany (www.micromata.com)
//
// ProjectForge is dual-licensed.
//
// This community edition is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License as published
// by the Free Software Foundation; version 3 of the License.
//
// This community edition is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
// Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, see http://www.gnu.org/licenses/.
//
/////////////////////////////////////////////////////////////////////////////

package org.projectforge.business.orga

import com.fasterxml.jackson.annotation.JsonIgnore
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.Indexed
import org.projectforge.common.anots.PropertyInfo
import org.projectforge.framework.jcr.AttachmentsInfo
import org.projectforge.framework.persistence.entities.DefaultBaseDO
import java.time.LocalDate
import jakarta.persistence.*
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.FullTextField
import org.hibernate.search.mapper.pojo.mapping.definition.annotation.GenericField
import org.projectforge.framework.persistence.history.NoHistory

/**
 * @author Kai Reinhard (k.reinhard@micromata.de)
 */
@Entity
@Indexed
@Table(name = "T_CONTRACT")
@NamedQueries(
        NamedQuery(name = ContractDO.FIND_OTHER_BY_NUMBER, query = "from ContractDO where number=:number and id<>:id"),
        NamedQuery(name = ContractDO.SELECT_MIN_MAX_DATE, query = "select min(date), max(date) from ContractDO"))
open class ContractDO : DefaultBaseDO(), AttachmentsInfo {

    @PropertyInfo(i18nKey = "'C-", additionalI18nKey = "legalAffaires.contract.number", tooltip = "fibu.tooltip.nummerWirdAutomatischVergeben")
    @GenericField // was: @FullTextField(analyze = Analyze.NO, bridge = FieldBridge(impl = IntegerBridge::class))
    @get:Column(name = "number")
    open var number: Int? = null

    @PropertyInfo(i18nKey = "date")
    @GenericField // was: @FullTextField(analyze = Analyze.NO)
    @get:Column(name = "c_date")
    open var date: LocalDate? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.validity.from")
    @GenericField // was: @FullTextField(analyze = Analyze.NO)
    @get:Column(name = "valid_from")
    open var validFrom: LocalDate? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.validity.until")
    @GenericField // was: @FullTextField(analyze = Analyze.NO)
    @get:Column(name = "valid_until")
    open var validUntil: LocalDate? = null

    @PropertyInfo(i18nKey = "title", required = true)
    @FullTextField
    @get:Column(length = 1000)
    open var title: String? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.coContractorA")
    @FullTextField
    @get:Column(length = 1000, name = "co_contractor_a")
    open var coContractorA: String? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.contractPersonA")
    @FullTextField
    @get:Column(length = 1000, name = "contract_person_a")
    open var contractPersonA: String? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.signerA")
    @FullTextField
    @get:Column(length = 1000, name = "signer_a")
    open var signerA: String? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.coContractorB")
    @FullTextField
    @get:Column(length = 1000, name = "co_contractor_b")
    open var coContractorB: String? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.contractPersonB")
    @FullTextField
    @get:Column(length = 1000, name = "contract_person_b")
    open var contractPersonB: String? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.signerB")
    @FullTextField
    @get:Column(length = 1000, name = "signer_b")
    open var signerB: String? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.signing")
    @GenericField // was: @FullTextField(analyze = Analyze.NO)
    @get:Column(name = "signing_date")
    open var signingDate: LocalDate? = null

    @PropertyInfo(i18nKey = "legalAffaires.contract.type")
    @FullTextField
    @get:Column(length = 100)
    open var type: String? = null

    @PropertyInfo(i18nKey = "status")
    @FullTextField
    @get:Enumerated(EnumType.STRING)
    @get:Column(length = 100)
    open var status: ContractStatus? = null

    @PropertyInfo(i18nKey = "text")
    @FullTextField
    @get:Column(length = 4000)
    open var text: String? = null

    @PropertyInfo(i18nKey = "fibu.common.reference")
    @FullTextField
    @get:Column
    open var reference: String? = null

    @PropertyInfo(i18nKey = "filing")
    @FullTextField
    @get:Column(length = 1000)
    open var filing: String? = null

    @PropertyInfo(i18nKey = "resubmissionOnDate")
    @GenericField // was: @FullTextField(analyze = Analyze.NO)
    @get:Column(name = "resubmission_on_date")
    open var resubmissionOnDate: LocalDate? = null

    @PropertyInfo(i18nKey = "dueDate")
    @GenericField // was: @FullTextField(analyze = Analyze.NO)
    @get:Column(name = "due_date")
    open var dueDate: LocalDate? = null

    @JsonIgnore
    @FullTextField
    @NoHistory
    @get:Column(length = 10000, name = "attachments_names")
    override var attachmentsNames: String? = null

    @JsonIgnore
    @FullTextField
    @NoHistory
    @get:Column(length = 10000, name = "attachments_ids")
    override var attachmentsIds: String? = null

    @JsonIgnore
    @NoHistory
    @get:Column(length = 10000, name = "attachments_counter")
    override var attachmentsCounter: Int? = null

    @JsonIgnore
    @NoHistory
    @get:Column(length = 10000, name = "attachments_size")
    override var attachmentsSize: Long? = null

    @PropertyInfo(i18nKey = "attachment")
    @JsonIgnore
    @get:Column(length = 10000, name = "attachments_last_user_action")
    override var attachmentsLastUserAction: String? = null

    companion object {
        internal const val FIND_OTHER_BY_NUMBER = "ContractDO_FindOtherByNumber"

        internal const val SELECT_MIN_MAX_DATE = "ContractDO_SelectMinMaxDate"
    }
}
